crimsobot/cogs/cringo.py

```diff
@@ -1,225 +1,423 @@
+    async def process_player_joining(self, player: discord.User) -> discord.Embed:
+        current_balance = await crimsogames.check_balance(player)  # cost is NOT debited.
+        not_enough_coin = current_balance < (self.minimum_balance if self.minimum_balance != 0 else float('-inf'))
+
+        # deny the poor of any opportunities
+        if not_enough_coin:
+            return self.generate_join_low_balance_embed(player)
+
+        # they're already playing!
+        if player.id in self.cog.all_players:
+            self.bounced.append(player)
+            return self.generate_join_already_playing_embed(player)
+
+        # see if we can send them messages, and if so, add them to the game
+        try:
+            embed = self.generate_join_test_message_embed()
+            await player.send(embed=embed)
+            self.joined.append(player)
+            self.cog.all_players.add(player.id)
+            return self.generate_join_success_embed(player)
+        except discord.errors.Forbidden:
+            self.bounced.append(player)
+            return self.generate_join_no_dms_embed(player)
+
+    # this function is called by the response handler
+    async def process_player_response(self, response: discord.Message) -> None:
+        try:  # find the player object of the response author
+            player = [player for player in self.players if player.user == response.author][0]
+        except IndexError:  # this player somehow isn't in the game? just ignore them
+            return
 
-            await ctx.send(embed=embed)
-            await asyncio.sleep(7)
+        # determine if user's response is a match
+        # matches missed in previous rounds are OK (they only lose the earlier round multiplier)
+        positions: List[str] = response.content.replace('.', '').strip().split(' ')
+        mismatch_detected = False
+        for position in positions:
+            if position == 'leave':  # evict them if they ask
+                embed = self.remove_from_game(player.user.id)
+                await self.context.send(embed=embed)
+                return
+
+            # if they're still in the game, then check for matches
+            match = await cringo.mark_card(player, position, self.used_emoji, self.multiplier)
+
+            if not match:
+                mismatch_detected = True
+
+        # mismatch message is sent only once
+        if mismatch_detected:
+            embed = self.generate_mismatch_embed()
+            await response.author.send(embed=embed)
+
+        await response.author.send(self.prettify_card(player.card))
 
-            # choose emojis, send to channel
-            emojis_this_turn = await cringo.cringo_emoji(1, card_size, emojis_already_used)
-            emojis_already_used.extend(emojis_this_turn[0])
-            multiplier = total_turns + 1 - turn
+    # this is where the majority of game logic is
+    async def start(self, ctx: commands.Context) -> None:
+        self.context = ctx
```

Comment:
I would move setting `ctx` to `__init__`, so you'd instead have `game = CringoGame(ctx)` and then nothing passed to `start()`